### PR TITLE
u32 bitwise imm value additional check

### PR DESCRIPTION
**Pull request by Overcastan**
u32 bitwise operations now have an additional check that the immediate value is 0 to prevent invalid values from being generated and unnecessary operations from being performed.
Related issue: #541

### CODE REVIEW

File: assembly/src/assembler/instruction/u32_ops.rs [hunk 1]
```diff
@@ -297,16 +294,21 @@ pub fn u32rotr(
...
-            // guaranteed to be u32 values
-            validate_param(imm, 0, MAX_U32_ROTATE_VALUE)?;
-            span.push_ops([Pad, U32assert2, Drop, Push(Felt::new(1 << (32 - imm)))]);
+            // if rotation is performed by 0, just verify that stack top is u32
+            if imm == 0 {
+                span.push_ops([Pad, U32assert2, Drop]);
+                return Ok(None);
+            }
+            validate_param(imm, 1, MAX_U32_ROTATE_VALUE)?;
+            span.push_ops([Push(Felt::new(1 << (32 - imm))), U32assert2]);
```
Comment (vlopes11) [inline assembly/src/assembler/instruction/u32_ops.rs:303]: Shouldn't we still perform the `Pad, U32assert2, Drop` if `imm != 0`?

File: assembly/src/assembler/instruction/u32_ops.rs [hunk 2]
```diff
@@ -297,16 +294,21 @@ pub fn u32rotr(
...
         }
         (Some(imm), U32OpMode::Unchecked) => {
-            validate_param(imm, 0, MAX_U32_ROTATE_VALUE)?;
+            // if rotation is performed by 0, do nothing (Noop)
```
Comment (vlopes11) [inline assembly/src/assembler/instruction/u32_ops.rs:306]: I think its more readable to create a dedicated variant on the match pattern instead of opening a new level

```rust
(Some(imm), U32OpMode::Unchecked) if imm == 0 => span.push_op(Noop),
(Some(imm), U32OpMode::Unchecked) => (..),
```

**Review by vlopes11 — CHANGES_REQUESTED** [review]

File: assembly/src/assembler/instruction/u32_ops.rs [hunk 3]
```diff
@@ -420,14 +422,24 @@ fn prepare_bitwise<const MAX_VALUE: u8>(
     span: &mut SpanBuilder,
     imm: Option<u8>,
     op_mode: U32OpMode,
-) -> Result<(), AssemblyError> {
+    ending_ops: [Operation; 2],
```
Comment (bobbinth) [inline assembly/src/assembler/instruction/u32_ops.rs:425]: minor nit: I'd call this `final_ops`.

File: assembly/src/assembler/instruction/u32_ops.rs (hunk 1 above)
Comment (bobbinth) [reply assembly/src/assembler/instruction/u32_ops.rs:303]: I think this is taken care of by the `U32assert2` online 303. Basically, instead of `Pad, U32assert2, Drop`, we do `push(value), U32assert2`.

**Comment by bobbinth** [comment]
Looks good! Thank you! I left a minor nit inline. But also one question: do we have tests for handling these edge cases, or should we add them?

File: assembly/src/assembler/instruction/u32_ops.rs (hunk 1 above)
Comment (Overcastan) [reply assembly/src/assembler/instruction/u32_ops.rs:303]: If I got it right, we were performing check like `Pad, U32assert2, Drop` because we wasn't sure that exponential value will fit u32 (so we checked only one value -- the one that we want to rotate, that's why we need `Pad` and `Drop`). After we check that `imm != 0` there is no possibility that exponential value won't fit u32, so we can check both values on the stack, and that's made by `U32assert2` after `Push` operation, so in that case we don't need `Pad` and `Drop` any more

**Comment by Overcastan** [comment]
Yes, we have tests for case when we rotate on 0. So it's working well, checked!

File: assembly/src/assembler/instruction/u32_ops.rs (hunk 1 above)
Comment (vlopes11) [reply assembly/src/assembler/instruction/u32_ops.rs:303]: Yes, that is correct! The comment is N/A

**Review by vlopes11 — APPROVED** [review]
Looks good! As @bobbinth mentioned, its always good to add tests for each PR, but this will not introduce new features. We might want to have some tool to check code coverage so we have a clear vision on when we are testing newly added behavior

**Review by bobbinth — APPROVED** [review]
All looks good to me! Thank you!
